ast/functions.go

```diff
@@ -382,3 +402,219 @@ func (n *AggregateFuncExpr) Accept(v Visitor) (Node, bool) {
+func (n *AggregateFuncExpr) Update() error {
+	name := strings.ToLower(n.F)
+	switch name {
+	case AggFuncCount:
+		return n.updateCount()
+	case AggFuncFirstRow:
+		return n.updateFirstRow()
+	}
+	return nil
+}
+
+// GetContext gets aggregate evaluation context for the current group.
+// If it is nil, add a new context into ContextPerGroup.
+func (n *AggregateFuncExpr) GetContext() *AggEvaluateContext {
+	if n.ContextPerGroup == nil {
+		n.ContextPerGroup = make(map[string](*AggEvaluateContext))
+	}
+	if _, ok := n.ContextPerGroup[n.CurrentGroup]; !ok {
+		c := &AggEvaluateContext{}
+		if n.Distinct {
+			c.Distinct = CreateAggregateDistinct(n.F, n.Distinct)
+		}
+		n.ContextPerGroup[n.CurrentGroup] = c
+	}
+	return n.ContextPerGroup[n.CurrentGroup]
+}
+
+func (n *AggregateFuncExpr) updateCount() error {
+	ctx := n.GetContext()
+	var value interface{}
+	if len(n.Args) > 0 {
+		hasNil := false
+		for _, a := range n.Args {
+			value = a.GetValue()
+			if value == nil {
+				hasNil = true
+			}
+		}
+		if hasNil {
+			return nil
+		}
+	}
+	if n.Distinct {
+		// TODO: compose values into value. For example select count(DISTINCT c1, c2) from t;
+		d, err := ctx.Distinct.IsDistinct(value)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !d {
+			return nil
+		}
+	}
+	ctx.Count++
+	return nil
+}
+
+func (n *AggregateFuncExpr) updateFirstRow() error {
+	ctx := n.GetContext()
+	if ctx.evaluated {
+		return nil
+	}
+	if len(n.Args) != 1 {
+		return errors.New("Wrong number of args for AggFuncFirstRow")
+	}
+	ctx.Value = n.Args[0].GetValue()
+	ctx.evaluated = true
+	return nil
+}
+
+// AggFuncDetector visits Expr tree to check if it contains AggregateFuncExpr.
+type AggFuncDetector struct {
+	HasAggFunc bool
+	detecting  bool
+}
+
+// Enter implements Visitor interface.
+func (a *AggFuncDetector) Enter(n Node) (node Node, skipChildren bool) {
+	defer func() {
+		a.detecting = true
+	}()
+	switch n.(type) {
+	case *AggregateFuncExpr, *GroupByClause:
+		a.HasAggFunc = true
+	case *SelectStmt, *InsertStmt, *DeleteStmt, *UpdateStmt:
+		if a.detecting {
+			// Enter a new context, skip it.
+			// For example: select sum(c) + c + exists(select c from t) from t;
+			return n, true
+		}
+	}
+	return n, a.HasAggFunc
+}
+
+// Leave implements Visitor interface.
+func (a *AggFuncDetector) Leave(n Node) (node Node, ok bool) {
+	return n, !a.HasAggFunc
+}
+
+// AggregateFuncExtractor visits Expr tree.
+// It converts ColunmNameExpr to AggregateFuncExpr and collects AggregateFuncExpr.
+type AggregateFuncExtractor struct {
+	inAggregateFuncExpr bool
+	// AggFuncs is the collected AggregateFuncExprs.
+	AggFuncs  []*AggregateFuncExpr
+	detecting bool
+}
+
+// Enter implements Visitor interface.
+func (a *AggregateFuncExtractor) Enter(n Node) (node Node, skipChildren bool) {
+	defer func() {
+		a.detecting = true
+	}()
+	switch v := n.(type) {
+	case *AggregateFuncExpr:
+		a.inAggregateFuncExpr = true
+		a.AggFuncs = append(a.AggFuncs, v)
```

Comment:
I think this line should be moved to `Leave`, all AggFunc will be appended in one place.